zilliqa/src/node.rs

```diff
@@ -263,26 +263,65 @@ impl Node {
         self.consensus.head_block().header.number
     }
 
-    pub fn resolve_block_number(&self, number: BlockNumberOrTag) -> u64 {
-        // TODO(#863): This doesn't respect two-chain finalization.
-        match number {
-            BlockNumberOrTag::Latest => self.get_chain_tip(),
-            BlockNumberOrTag::Finalized => self.get_chain_tip().saturating_sub(2),
-            BlockNumberOrTag::Safe => self.get_chain_tip().saturating_sub(2),
+    pub fn resolve_block_number(&self, block_number: BlockNumberOrTag) -> u64 {
+        match block_number {
+            BlockNumberOrTag::Number(n) => n,
             BlockNumberOrTag::Earliest => 0,
+            BlockNumberOrTag::Latest => self.get_chain_tip(),
             BlockNumberOrTag::Pending => self.get_chain_tip(),
-            BlockNumberOrTag::Number(n) => n,
+            BlockNumberOrTag::Finalized => {
+                let Ok(Some(view)) = self.db.get_latest_finalized_view() else {
+                    return 0u64;
+                };
+                let Ok(Some(block)) = self.db.get_block_by_view(&view) else {
+                    return 0u64;
+                };
+                block.number()
+            }
+            // From whitepaper: If the proposed block’s view number
+            // is one larger than its QC’s view number and
+            // is larger or equal to the validator’s local
+            // view number, the validator regards the
+            // proposed block as safe (which, in the codebase, is referenced as head_block())
```

Comment:
I'm afraid the terminology used in the HotStuff papers I adhere to in the whitepaper caused a misunderstanding. The term `safe` in HotStuff has nothing to do with the `safe` block tag in the eth APIs.

On post-Merge Ethereum, the `safe` block corresponds to a justified checkpoint: https://github.com/ethereum/consensus-specs/blob/v1.3.0/fork_choice/safe-block.md

The term `justified` comes from Tendermint, and it means it has received 2/3 of the validator votes. Note that the term `checkpoint` here has nothing to do with the weak-subjectivity checkpoints Theo is working on. Ethereum FFG (which is based on the Tendermint algorithm) uses the term `checkpoint` to refer to the first slot of an epoch, which is a candidate for justification (1. voting round) and finalization (2. voting round).

Long story short, the `safe` block tag in our consensus refers to the block that the node's highQC points to (highQC means it's the latest = high, and it's a QC = 2/3 validators voted for it).